internal/api/option/error_test.go

```diff
@@ -0,0 +1,120 @@
+	svcv1alpha1connect.UnimplementedKargoServiceHandler
+	getVersionInfoHandler getVersionInfoHandlerFunc
+}
+
+func newTestErrorInterceptorServer(
+	getVersionInfoHandler getVersionInfoHandlerFunc,
+) *testErrorInterceptorServer {
+	return &testErrorInterceptorServer{
+		getVersionInfoHandler: getVersionInfoHandler,
+	}
+}
+
+func (s *testErrorInterceptorServer) GetVersionInfo(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.GetVersionInfoRequest],
+) (*connect.Response[svcv1alpha1.GetVersionInfoResponse], error) {
+	return s.getVersionInfoHandler(ctx, req)
+}
+
+func TestErrorInterceptor(t *testing.T) {
+	testCases := map[string]struct {
+		handlerFunc        getVersionInfoHandlerFunc
+		errExpected        bool
+		expectedStatusCode connect.Code
+	}{
+		"no error": {
+			handlerFunc: func(
+				context.Context,
+				*connect.Request[svcv1alpha1.GetVersionInfoRequest],
+			) (*connect.Response[svcv1alpha1.GetVersionInfoResponse], error) {
+				return connect.NewResponse(&svcv1alpha1.GetVersionInfoResponse{}), nil
+			},
+			errExpected: false,
+		},
+		"interceptor should not unwrap connect error with explicit status code": {
+			handlerFunc: func(
+				context.Context,
+				*connect.Request[svcv1alpha1.GetVersionInfoRequest],
+			) (*connect.Response[svcv1alpha1.GetVersionInfoResponse], error) {
+				return nil, connect.NewError(
+					connect.CodeInternal,
+					kubeerr.NewForbidden(schema.GroupResource{}, "", nil),
+				)
+			},
+			errExpected:        true,
+			expectedStatusCode: connect.CodeInternal,
+		},
+		"interceptor should not unwrap connect error with unknown status code": {
+			handlerFunc: func(
+				context.Context,
+				*connect.Request[svcv1alpha1.GetVersionInfoRequest],
+			) (*connect.Response[svcv1alpha1.GetVersionInfoResponse], error) {
+				return nil, connect.NewError(
+					connect.CodeUnknown,
+					kubeerr.NewForbidden(schema.GroupResource{}, "", nil),
+				)
+			},
+			errExpected:        true,
+			expectedStatusCode: connect.CodeUnknown,
+		},
+		"interceptor should wrap error with appropriate status code if possible": {
+			handlerFunc: func(
+				context.Context,
+				*connect.Request[svcv1alpha1.GetVersionInfoRequest],
+			) (*connect.Response[svcv1alpha1.GetVersionInfoResponse], error) {
+				return nil, kubeerr.NewForbidden(schema.GroupResource{}, "", nil)
+			},
+			errExpected:        true,
+			expectedStatusCode: connect.CodePermissionDenied,
+		},
+	}
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			opt := connect.WithInterceptors(newErrorInterceptor())
+			mux := http.NewServeMux()
+			handler := newTestErrorInterceptorServer(tc.handlerFunc)
+			mux.Handle(svcv1alpha1connect.NewKargoServiceHandler(handler, opt))
+			srv := httptest.NewServer(mux)
+			srv.EnableHTTP2 = true
+			t.Cleanup(srv.Close)
+
+			cli := svcv1alpha1connect.NewKargoServiceClient(srv.Client(), srv.URL, connect.WithGRPC())
+			_, err := cli.GetVersionInfo(
```

Comment:
Why don't we just call `wrapError()` directly to see what it does for different `tc.err`?

Reply:
Since the subject of test is `interceptor`, and I think test code doesn't have to know about implementation details, it would be better to test interceptor flow to make sure it works.

Reply:
> Since the subject of test is interceptor

With a majority of the logic being in the `toConnectError()` function (formerly `wrapError()`), it seems that if the function were viewed as the test subject instead of the entire type, validating the important logic would be accomplished more simply.

_I'll concede that you squeeze out slightly better test coverage with your approach, so I won't die on this hill._

> I think test code doesn't have to know about implementation details

Kargo uses whitebox testing pretty extensively. Knowledge of the code under test often leads to better coverage of edge cases.